src/filter/opt.rs

```diff
@@ -8,6 +8,8 @@ use super::*;
 lazy_static! {
     static ref OPTIMIZED: std::sync::Mutex<std::collections::HashMap<Filter, Filter>> =
         std::sync::Mutex::new(std::collections::HashMap::new());
+    static ref INVERTED: std::sync::Mutex<std::collections::HashMap<Filter, Filter>> =
```

Comment:
For complex filters this can be a somewhat expensive recursive call, and it is done repeatedly in a few places.
I changed it though to only cache the recursive calls and not the simple cases now.